src/registries/npm-registry.js

```diff
@@ -55,7 +55,10 @@ export default class NpmRegistry extends Registry {
 
   request(pathname: string, opts?: RegistryRequestOptions = {}, packageName: ?string): Promise<*> {
     const registry = this.getRegistry(packageName || pathname);
-    const requestUrl = url.resolve(registry, pathname);
+    const requestBaseUrl = url.resolve(registry, pathname); // does not work...
+    const requestUrl = requestBaseUrl.replace('https://registry.yarnpkg.com/', registry);
```

Comment:
I think you are on a right path.
Just make the domains configurable via .yarnrc.
Those can be read in config via getOption, e.g. https://github.com/yarnpkg/yarn/blob/master/src/config.js#L267